### DISCUSSION TITLE
Experimental USB support - ESP32 S3

### DISCUSSION THREAD

**Discussion by mkellner** [Moddable-OpenSource/moddable#971] [General]
## Experimental USB support

For the esp32s3, Moddable has added experimental support for programming and debugging over the USB port using a MacOS host.

### ESP32-S3-DevKitC-1-N8

The Espressif ESP32-S3-DevKitC-1-N8 has two USB ports on the board. One port is connected to a USB-to-serial chip and allows the device to be programmed and debugged like other ESP32 devices. Use the platform `esp32/esp32s3` for this configuration.

The other port is connected to the USB pins on the ESP32-S3. Use the platform `esp32/esp32s3_usb` for this configuration.

### Adafruit QT Py ESP32-S3

The Adafruit QT Py ESP32-S3 has one USB port connected to the USB pins on the ESP32-S3 and no built-in USB-to-serial. Use the platform `esp32/qtpys3` for this device.

## Programming over USB

To program over USB, put the device in programming mode by pressing and holding the boot button while pressing and releasing the reset button. Then release the boot button.

Ensure the `UPLOAD_PORT` environment variable is not set:

```
unset UPLOAD_PORT
```

Build and install by using the `mcconfig` command like:

```
cd $MODDABLE/examples/helloworld
mcconfig -d -m -p esp32/qtpys3
```

After the install is done, the terminal will show:

```
Executing "ninja flash"...
Done
```

Press the `reset` button on the device. The terminal will show something like:

```
product/vendor match: /dev/cu.usbmodem1234561
```

and the device will connect to xsbug.

**Comment by MKGaru** [comment]
Awesome! I test on Banana Pi BPI-Leaf-S3 with use target ( esp32/esp32s3_usb ) . It' works fine.
https://wiki.banana-pi.org/BPI-Leaf-S3
